xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/js/xwiki/table/livetable.js

```diff
@@ -303,18 +303,22 @@ XWiki.widgets.LiveTable = Class.create({
 
     for (var i = off; i <= f; i++) {
       if (this.fetchedRows[i]) {
-        var elem = this.handler(this.fetchedRows[i], i, this);
+        var elem = this.fetchedRows[i].livetableRenderedRow;
+        if (elem === undefined) {
+          this.fetchedRows[i].livetableRenderedRow = this.handler(this.fetchedRows[i], i, this);
+          elem = this.fetchedRows[i].livetableRenderedRow;
+          var memo = {
+            "data": this.fetchedRows[i],
+            "row":elem,
+            "table":this,
+            "tableId":this.domNodeName
+          };
+          // 1. Named event (for code interested by that table only)
+          document.fire("xwiki:livetable:" + this.domNodeName + ":newrow", memo);
+          // 2. Generic event (for code potentially interested in any livetable)
+          document.fire("xwiki:livetable:newrow", memo);
```

Comment:
Note that this will slightly change the meaning of the `newrow` event, in the way that it will no longer trigger on each **display** of a livetable row, but on the actual **rendering** of a new livetable row, which is done only once for the set of cached data.

In practice, I believe this change will have no negative impact, as the general usecase for this event is to augment the rendering of a livetable row and, since we will cache each rendered row, it will be enough (and better performance-wise) to do it once and reuse it on each display.